data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java

```diff
@@ -49,36 +52,75 @@ public Collection<Record<Event>> doExecute(final Collection<Record<Event>> recor
             }
 
             try {
-                final Map<String, Object> sourceMap = recordEvent.get(config.getSource(), Map.class);
-                final List<Map<String, Object>> targetList = new ArrayList<>();
-
-                Map<String, Object> modifiedSourceMap = new HashMap<>();
-                for (final Map.Entry<String, Object> entry : sourceMap.entrySet()) {
-                    if (excludeKeySet.contains(entry.getKey())) {
-                        if (config.getRemoveProcessedFields()) {
-                            modifiedSourceMap.put(entry.getKey(), entry.getValue());
+                final Map<String, Object> sourceMap = getSourceMap(recordEvent);
+
+                if (config.getConvertFieldToList()) {
+                    final List<List<Object>> targetNestedList = new ArrayList<>();
+
+                    for (final Map.Entry<String, Object> entry : sourceMap.entrySet()) {
+                        if (!excludeKeySet.contains(entry.getKey())) {
+                            targetNestedList.add(List.of(entry.getKey(), entry.getValue()));
                         }
-                        continue;
-                    }
-                    targetList.add(Map.of(
-                            config.getKeyName(), entry.getKey(),
-                            config.getValueName(), entry.getValue()
-                    ));
-                }
 
-                if (config.getRemoveProcessedFields()) {
-                    recordEvent.put(config.getSource(), modifiedSourceMap);
+                    }
+                    removeProcessedFields(sourceMap, recordEvent);
+                    recordEvent.put(config.getTarget(), targetNestedList);
+                } else {
+                    final List<Map<String, Object>> targetList = new ArrayList<>();
+                    for (final Map.Entry<String, Object> entry : sourceMap.entrySet()) {
+                        if (!excludeKeySet.contains(entry.getKey())) {
+                            targetList.add(Map.of(
+                                    config.getKeyName(), entry.getKey(),
+                                    config.getValueName(), entry.getValue()
+                            ));
+                        }
+                    }
+                    removeProcessedFields(sourceMap, recordEvent);
+                    recordEvent.put(config.getTarget(), targetList);
                 }
-
-                recordEvent.put(config.getTarget(), targetList);
             } catch (Exception e) {
                 LOG.error("Fail to perform Map to List operation", e);
-                //TODO: add tagging on failure
+                recordEvent.getMetadata().addTags(config.getTagsOnFailure());
             }
         }
         return records;
     }
 
+    private Map<String, Object> getSourceMap(Event recordEvent) throws JsonProcessingException {
+        final Map<String, Object> sourceMap;
+        if (config.getSource() == null) {
```

Comment:
Is this a desirable feature?

If so, can we make it take in `/` instead of `null`? I think this could lead to confusion when users don't set the `source` field they are probably not getting what they want.

Reply:
Yes, we will need it for #3965. 

> If so, can we make it take in / instead of null? I think this could lead to confusion when users don't set the source field they are probably not getting what they want.

In [JsonPointer specification](https://datatracker.ietf.org/doc/html/draft-ietf-appsawg-json-pointer-03#section-5), it is actually the empty string `""` that refers to the whole document and `"/"` refers to the field with `""` as key. If we remove the restrictions in JacksonEvent that key cannot be empty, then we should already support referring to the root document with Json Pointer `""`. What do you think?

Reply:
@oeyh , That is a very useful finding.

I think we will need to make the syntax be explicit then:

```
source: ''
```

Reply:
Implemented this in recent commits.